### PR TITLE
document: identifiers management

### PR DESCRIPTION
**Pull request by zannkukai**
* Add classes and methods to have a best document identifiers management.
* Add a behavior to render an identifier with a specific format
  (depending of the desired output purpose).

Co-Authored-by: Renaud Michotte <renaud.michotte@gmail.com>
Co-Authored-by: Bertrand Zuchuat <bertrand.zuchuat@rero.ch>

## Code review check list

- [ ] Commit message template compliance.
- [ ] Commit message without typos.
- [ ] File names.
- [ ] Functions names.
- [ ] Functions docstrings.
- [ ] Unnecessary commited files?
- [ ] Cypress tests successful?

### CODE REVIEW

**Comment by zannkukai** [comment]
This structure will be used in next commits to manage document identifiers (search, indexing, ...)

**Review by lauren-d — APPROVED** [review]

**Review by BadrAly — APPROVED** [review]

File: rero_ils/modules/documents/commons/identifiers.py [hunk 1]
```diff
@@ -0,0 +1,281 @@
+# -*- coding: utf-8 -*-
```
Comment (jma) [inline rero_ils/modules/documents/commons/identifiers.py:1]: Identifiers can be used by other type of resources such as contributions. Maybe, you can put it at the modules level?

File: rero_ils/modules/documents/commons/identifiers.py [hunk 2]
```diff
@@ -0,0 +1,281 @@
...
+    MUSIC_PLATE = 'bf:MusicPlate'
+    MUSIC_PUBLISHER_NUMBER = 'bf:MusicPublisherNumber'
+    PUBLISHER_NUMBER = 'bf:PublisherNumber'
+    UPC = 'bf:Upc'
+    URN = 'bf:Urn'
+    VIDEO_RECORDING_NUMBER = 'bf:VideoRecordingNumber'
+    URI = 'uri'
+
+
+class IdentifierStatus:
```
Comment (jma) [inline rero_ils/modules/documents/commons/identifiers.py:85]: I do not understand what is the purpose of this. You can perhaps use the status of the invenio persistent identifiers:
https://github.com/inveniosoftware/invenio-pidstore/blob/e58e0987e7bcc8d8b021f44a88b401b89b8ae8c8/invenio_pidstore/models.py#L35

Comment (jma) [inline rero_ils/modules/documents/commons/identifiers.py:85]: I do not understand what is the purpose of this. You can perhaps use the status of the invenio persistent identifiers:
https://github.com/inveniosoftware/invenio-pidstore/blob/e58e0987e7bcc8d8b021f44a88b401b89b8ae8c8/invenio_pidstore/models.py#L35

File: rero_ils/modules/documents/commons/identifiers.py [hunk 3]
```diff
@@ -0,0 +1,281 @@
...
+            raise InvalidIdentifierException()
+
+    def render(self, render_class=None) -> str:
+        """Render the identifier.
+
+        :param render_class: a rendering class to use to get the identifier
+                             representation.
+        :returns the string representation of the identifier.
+        """
+        render_class = render_class or DefaultIdentifierRenderer()
```
Comment (jma) [inline rero_ils/modules/documents/commons/identifiers.py:155]: I would use a class variable for this as done for the model_cls in the record class.

Comment (jma) [inline rero_ils/modules/documents/commons/identifiers.py:155]: I would use a class variable for this as done for the model_cls in the record class.

File: rero_ils/modules/documents/commons/identifiers.py [hunk 4]
```diff
@@ -0,0 +1,281 @@
...
+        return True
+
+    def validate(self) -> None:
+        """Validate the identifier."""
+        # same as `is_valid` but raise an exception if the identifier isn't
+        # valid
+        if not self.is_valid():
+            raise InvalidIdentifierException()
+
+    def render(self, render_class=None) -> str:
```
Comment (jma) [inline rero_ils/modules/documents/commons/identifiers.py:148]: I would add a `**kwargs` to pass to the render class.

Comment (jma) [inline rero_ils/modules/documents/commons/identifiers.py:148]: I would add a `**kwargs` to pass to the render class.

**Review by jma — CHANGES_REQUESTED** [review]

File: rero_ils/modules/documents/commons/identifiers.py (hunk 2 above)
Comment (zannkukai) [reply rero_ils/modules/documents/commons/identifiers.py:85]: I'm in the same feeling :-) Maybe it is an old information from Virtua ? 
In the document `identifiedBy`, there is a "status" field : https://github.com/rero/rero-ils/blob/staging/rero_ils/modules/documents/jsonschemas/documents/document_identified_by-v0.0.1.json#L199-L214

File: rero_ils/modules/documents/commons/identifiers.py (hunk 3 above)
Comment (zannkukai) [reply rero_ils/modules/documents/commons/identifiers.py:155]: I don't think so. This is not a constant as `model_cls`. In this case we can ask to render the same identifier in a different way :
```python
data = {
    'type': 'bf:Isbn', 
    'value': '9781234567890', 
    'qualifier': 'tome 2', 
    'source': 'RERO'
}
identifier = IdentifierFactory.create_identifier(data)
print(identifier.render()) # As we don't use render_class parameter, we use the DefaultIdentifierRenderer
>>> 9781234567890
print(identifier.render(render_class=QualifierIdentifierRenderer()))
>>> 9781234567890, tome 2
print(identifier.render(render_class=NormalizedIdentifierRenderer()))
>>> 978-1-2345678-90
print(identifier.render(render_class=NormalizedIdentifierWithSourceRenderer()))
>>> [RERO] 978-1-2345678-90
...
```

File: rero_ils/modules/documents/commons/identifiers.py (hunk 4 above)
Comment (zannkukai) [reply rero_ils/modules/documents/commons/identifiers.py:148]: done

File: rero_ils/modules/documents/commons/identifiers.py (hunk 1 above)
Comment (zannkukai) [reply rero_ils/modules/documents/commons/identifiers.py]: Moved to `rero_ils.modules.commons.py`

**Review by jma — APPROVED** [review]

File: rero_ils/modules/commons.py [hunk 5]
```diff
@@ -0,0 +1,282 @@
...
+
+
+class QualifierIdentifierRenderer(IdentifierRenderer):
+    """Identifier renderer class including the qualifier attribute."""
+
+    def render(self, identifier: Identifier, **kwargs) -> str:
+        """Get the string representation of an identifier."""
+        output = str(identifier)
+        if identifier.qualifier:
+            output += f', {identifier.qualifier}'
```
Comment (rerowep) [inline rero_ils/modules/commons.py:254]: Maybe clearer to write:
```python
output = f'{output}, {identifier.qualifier}'
```

**Pull request merged**